Output Link to ticket in found address CSV

gbirke opened the pull request:
The output CSV can now contain a ticket ID.
Move CSV generation to output class.
Delete CSVSourceDataReader - exported CSV has no ticket id and we're
using the database now.

gbirke committed:
Output Link to ticket in found address CSV

The output CSV can now contain a ticket ID.
Move CSV generation to output class.
Delete CSVSourceDataReader - exported CSV has no ticket id and we're
using the database now.

JeroenDeDauw commented on src/UseCases/ExtractAddress/CSVFormatter.php:
```diff
@@ -0,0 +1,17 @@
+
+trait CSVFormatter {
+
+	protected function formatAsCSV( array $row ): string {
```
Why not private?

JeroenDeDauw commented on src/UseCases/ExtractAddress/CSVFormatter.php:
```diff
@@ -0,0 +1,17 @@
+
+namespace WMDE\OtrsExtractAddress\UseCases\ExtractAddress;
+
+trait CSVFormatter {
```
Why trait and not composition? If you don't want to inject the thing I'd still make a class and just instantiate it where it is used.

JeroenDeDauw commented on src/UseCases/ExtractAddress/CSVFormatter.php:
```diff
@@ -0,0 +1,17 @@
+			if ( $v ) {
+				return '"'. str_replace( '"', '""', $v ) . '"';
+			} else {
+				return '';
```
Oh noez! an `if-esle`. You can use a guard followed by a return

JeroenDeDauw commented on src/UseCases/ExtractAddress/RejectedAddressWriter.php:
```diff
@@ -0,0 +1,33 @@
+
+	private $outputStream;
+
+	public function __construct( $outputStream ) {
```
Is it still not possible to type hint this in PHP 7.1?

gbirke replied:
Nope, resources (DB resource, file handler resource), can't be type hinted yet. But I made another commit where I replace the resource with `SplFileObject`.

JeroenDeDauw reviewed (commented):
This talk is really boring... so you can has some reviews :)

gbirke committed:
Refactor CSVFormatter

Convert trait to class and use that class internally.
Remove superfluous `else`.

JeroenDeDauw commented:
+0.82

gbirke committed:
Convert stream resources to SplFileObjects

Resources can't be type hinted yet in PHP (at least until 7.1), see
http://stackoverflow.com/questions/40637939/typehinting-for-filehandle-as-argument-in-function

So we use SplFileObject instead.

JeroenDeDauw commented on src/UseCases/ExtractAddress/RejectedAddressWriter.php:
```diff
@@ -14,14 +14,14 @@ class RejectedAddressWriter {
 	public function writeRow( SourceData $data, string $errorMsg ) {
 		// TODO add option to skip "not found" messages and export only other messages
-		fwrite( $this->outputStream, $this->csvFormatter->formatAsCSV( [
+		$this->outputStream->fwrite( $this->csvFormatter->formatAsCSV( [
```
cool... no more global call

JeroenDeDauw commented:
I don't see any obvious issues but cannot +2 without understanding the tasks and tools involved here more. Presumably @KaiNissen is better up to speed?
